kv/txn.go

```diff
@@ -17,23 +17,112 @@ package kv
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
 	"math/rand"
+	"sync"
 	"time"
 
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/parser/terror"
 	"github.com/pingcap/tidb/util/logutil"
+	"github.com/tikv/client-go/v2/oracle"
 	"go.uber.org/zap"
 )
 
+const (
+	// TimeToPrintLongTimeInternalTxn is the duration if the internal transaction lasts more than it,
+	// TiDB prints a log message
+	TimeToPrintLongTimeInternalTxn = time.Minute * 5
+)
+
+var globalInnerTxnTsBox = innerTxnStartTsBox{
+	innerTSLock:        sync.Mutex{},
+	innerTxnStartTsMap: make(map[uint64]struct{}, 256),
+}
+
+type innerTxnStartTsBox struct {
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]struct{}
+}
+
+// wrapStoreInterTxnTS is the entry function used to store the startTS of an internal transaction to globalInnerTxnTsBox
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapStoreInterTxnTS(startTS uint64) {
+	globalInnerTxnTsBox.storeInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) storeInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	ib.innerTxnStartTsMap[startTS] = struct{}{}
+	ib.innerTSLock.Unlock()
+
+}
+
+// wrapDeleteInterTxnTS delete the startTS from globalInnerTxnTsBox when the transaciotn is commted
+// @param startTS	the startTS of the internal transaction produced by `RunInNewTxn`
+func wrapDeleteInterTxnTS(startTS uint64) {
+	globalInnerTxnTsBox.deleteInnerTxnTS(startTS)
+}
+
+func (ib *innerTxnStartTsBox) deleteInnerTxnTS(startTS uint64) {
+	ib.innerTSLock.Lock()
+	delete(ib.innerTxnStartTsMap, startTS)
+	ib.innerTSLock.Unlock()
+}
+
+// WrapGetMinStartTs get the min StartTS between startTSLowerLimit and curMinStartTS in globalInnerTxnTsBox
+func WrapGetMinStartTs(now time.Time, startTSLowerLimit uint64,
+	curMinStartTS uint64) uint64 {
+	return globalInnerTxnTsBox.getMinStartTs(now, startTSLowerLimit, curMinStartTS)
+}
```

Comment:
Can we just make `globalInnerTxnTsBox` and `getMinStartTs` public because the wrappers are too simple?

Reply:
I agree

Reply:
My confusion is that I cannot understand what `Wrap` means here... So, it's actually hard for me to recognize what `WrapGetMinStartTS` is from the caller...
If you call transactions from `RunInNewTxn` as `InnerTxn`, what about renaming `WrapGetMinStartTs` to `GetMinInnerTxnStartTS`>

Reply:
ok